yellowbox/extras/logstash.py

```diff
@@ -1,30 +1,263 @@
+                if key.fileobj is self._rshutdown:
+                    assert self._rshutdown.recv(
+                        len(_CLOSE_SENTINEL)) == _CLOSE_SENTINEL
+                    return
+
+                # Handle new connection
+                if key.fileobj is self._root:
+                    new_socket, _ = self._root.accept()
+                    self._selector.register(new_socket, selectors.EVENT_READ,
+                                            self._create_data_callback(new_socket))
+                    continue
+
+                # Data received, run callback
+                try:
+                    key.data()
+                except Exception:
+                    _logger.exception("Unknown error occurred, closing connection.")
+                    self._selector.unregister(key.fileobj)
+                    # noinspection PyUnresolvedReferences
+                    key.fileobj.close()
+
+    def start(self, *, retry_spec: None = None) -> None:
+        """Start the service.
+
+        Args:
+            retry_spec: Ignored.
+        """
+        self._root.listen(_LISTEN_BACKLOG)
+        self._selector.register(self._root, selectors.EVENT_READ)
+        self._selector.register(self._rshutdown, selectors.EVENT_READ)
+        self._thread.start()
+        super(LogstashService, self).start(retry_spec=retry_spec)
+
+    def stop(self) -> None:
+        """Stop the service
+
+        Raises:
+            RuntimeError: Failed stopping the service. Operation timed out.
+        """
+        if not self._thread.is_alive():
+            return
+        self._wshutdown.send(_CLOSE_SENTINEL)
+        self._thread.join(_STOP_TIMEOUT)  # Should almost never timeout
+        if self._thread.is_alive():
+            raise RuntimeError(f"Failed stopping {self.__class__.__name__}.")
+
+        for key in self._selector.get_map().values():
+            sock = cast(socket.socket, key.fileobj)
+            sock.close()
+
+        self._selector.close()
+
+    def is_alive(self) -> bool:
+        """Check if LogstashService is alive.
+
+        Returns:
+            Boolean.
+        """
+        return self._thread.is_alive()
+
+    def connect(self, network: Any) -> List[str]:
+        """Does nothing. Conforms to YellowService interface."""
+        # Logstash service is not docker related. It cannot actually connect to
+        # the network. However, other containers connected to the network can
+        # connect to the service with docker's usual host
+        return [_docker_host_name]
+
+    def disconnect(self, network: Any):
+        """Does nothing. Conforms to YellowService interface."""
+        pass
+
+    def _filter_records(self, level: int) -> Iterable[Dict[str, Any]]:
+        """Filter records in the given level or above."""
+        return (record for record in self.records if
+                logging.getLevelName(record["level"]) >= level)
+
+    def assert_logs(self, level: Union[str, int]):
```

Comment:
this function will 100% cause bugs. It's not enough to assert that a message was received, the message should also conform to some other constraints. Either support these constraints or remove this method entirely.

Reply:
That's how unittest assets logs, I get your point but it never caused a bug for me. @aviramha what do you think?

Reply:
I agree with ben

Reply:
Which constraints?

Reply:
Doesn't matter how unitest does things, we aren't trying to emulate them. I want to ensure we receive a log with specific words, perhaps with specific tags.

Reply:
Specific tags is rather useless at least right now, as the tags are always an empty list in our systems. You can't even set them if you want to (aiologstash doesn't allow it).

Regarding specific text, I can add it if you guys want. I'm personally not going to use it but I can see a use for it.

Reply:
I think I have an idea for common ground - return the logs that match the given criteria (level) so it's basically the same as unit test but it lets the user easily retrieve logs in a context.